compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp

```diff
@@ -0,0 +1,369 @@
+    for (uint32_t i = 0; i < _back_transpose->rank(); i++)
+    {
+      shape_node->at<loco::DataType::S32>(i) = _back_transpose->dim(i).value();
+    }
+    shape_node->shape_status(luci::ShapeStatus::VALID);
+    shape_node->name(pattern_name + "/shape");
+  }
+
+  auto reshape_node = graph->nodes()->create<luci::CircleReshape>();
+  {
+    reshape_node->name(pattern_name);
+    reshape_node->tensor(_front_transpose->a());
+    reshape_node->shape(shape_node);
+    luci::add_origin(reshape_node, luci::get_origin(_mid_reshape));
+  }
+
+  return reshape_node;
+}
+
+bool remove_no_effect_reshape(luci::CircleTranspose *node)
+{
+  // find 'front_transpose - mid_reshape(reduce_dimension) - back_transpose' pattern
+  const auto back_transpose = node;
+  const auto mid_reshape = dynamic_cast<luci::CircleReshape *>(back_transpose->a());
+  {
+    if (mid_reshape == nullptr)
+      return false;
+
+    const auto shape_node = dynamic_cast<luci::CircleConst *>(mid_reshape->shape());
+    if (shape_node == nullptr)
+      return false;
+
+    if (shape_node->dtype() != loco::DataType::S32)
+      return false;
+
+    const auto input_rank = mid_reshape->rank();
+    const auto output_rank = shape_node->size<loco::DataType::S32>();
+
+    // for now, handle only rank reduction case
+    if (input_rank >= output_rank)
+      return false;
+  }
+  const auto front_transpose = dynamic_cast<luci::CircleTranspose *>(mid_reshape->tensor());
+  {
+    if (not front_transpose)
+      return false;
+  }
+
+  // simulate
+  PatternSimulator sim(front_transpose, mid_reshape, back_transpose);
+
+  sim.simulate();
+
+  if (sim.canRemoveTransposes() == false)
+    return false;
+
+  // replace
+  auto new_reshape_node = sim.createAlternative(node->graph());
+  replace(node).with(new_reshape_node);
+
+  return true;
+}
+
+} // namespace
+
+namespace luci
+{
+
+/**
+ * BEFORE
+ *
+ * - Target only 'in.rank() >= out.rank()'
+ *
+ *     [CircleNode]      [CircleConst]
+ *    (in)               (perm)
+ *         \              /
+ *       [CircleTranspose]    [CircleConst]
+ *            \               (shape)
+ *             \             /
+ *             [CircleReshape]      [CircleConst]
+ *               \                  (perm)
+ *                \                /
+ *                 [CircleTranspose]
+ *                  \
+ *                   \
+ *                    [CircleNode]
+ *                    (out)
+ *
+ * AFTER
+ *
+ *    [CircleNode]        [CircleConst]
+ *     (in)                (shape)
+ *        \               /
+ *        [CircleReshape]
+ *         (new)
+ *          \
+ *          [CircleNode]
+ *          (out)
+ *
+ */
+
+bool RemoveUnnecessaryTransposeNetPass::run(loco::Graph *g)
+{
+  bool changed = false;
+
+  for (auto node : loco::active_nodes(loco::output_nodes(g)))
+  {
+    if (auto transpose_node = dynamic_cast<luci::CircleTranspose *>(node))
+    {
+      if (remove_no_effect_reshape(transpose_node))
```

Comment:
Will be renamed, Sorry for confusion.

```suggestion
      if (remove_no_effect_transpose(transpose_node))
```